### PR TITLE
Improve performances in fetching data

### LINKED ISSUE #27: Improve performances in fetching a large datasets online
Using the erddap data source to fetch a large amount of data has poor performances (in terms of wall time).
Such kind of fetching is problematic because:
- the erddap has a 1 minute time out that will throw a bad response if it takes too long to prepare to output on the server side (see #16).
- the erddap server can only handle requests leading to no more than 9308MB of data.

To check this out, we can use the example from #16:
```python
from argopy import DataFetcher as ArgoDataFetcher
box = [-120., -85., -10, 10, 0, 500, '2019-01-01','2019-12-31']
ds = ArgoDataFetcher(src='erddap').region(box).to_xarray()
```

To fix this issue, and more importantly improve performances and reliability of large datasets fetching, I propose to test the following pattern:

In argopy.data_fetchers.ErddapArgoDataFetcher, implement a chunking procedure for http requests:
- Given a ``box`` request: split the domain into chunks/sub-domains: in space or time, or may be both. 
- Given a ``wmo`` request: split the request into chunks, one each wmo.

Each chunk of the full request would be fetched using its own http request and chunk data would be gathered/concatenated as they come along from the server.

Chunking the request, i.e. creating a list of urls to fetch, would be done in at the fetcher level in ``argopy.data_fetchers.ErddapArgoDataFetcher``.
Managing a pull of requests, would be done at the file system level in ``argopy.stores.httpstore``.

I suspect [zarr](https://github.com/zarr-developers/zarr-python) or [fsspec](https://github.com/intake/filesystem_spec) can already manage an (asynchronous) pull of requests.
In zarr, we can read it is capable of "Read an array concurrently from multiple threads or processes.", so this looks very close to the pattern above. May be @rabernat can give us his impressions on this ?

### PR DESCRIPTION
**Pull request by gmaze**
Add an asynchrone open_mfdataset to the httpstore
and much more ...

Close #27 #16 #51

### CODE REVIEW

**Comment by gmaze** [comment]
Here are some benchmarking results for the **time of data fetching using the erddap and a large box.**
Box limit varies to increase dataset to fetch in size.

The following code was repeated for box depth of 0-50m and 2018 alone and then 0-50m, 0-100m, 0-200m and 2018/2019.
The horizontal domain size is 50deg in longitude and 20deg in latitude.
I use small random variations around the longitude boundaries to avoid the erddap server to cache requests.
I also use the ``expert`` mode to limit the amount of postprocessing on the client side.
Note the new fetcher options ``parallel=True, chunks='auto', box_maxsize=[10, 10, 50]``.
We run the same request n=5 times to get an ensemble approach and more stable metrics.

```python
for run in range(5):
    start_time = time.time()
    large_box = [-70+np.random.random_sample(1)[0], -30+np.random.random_sample(1)[0], 20, 40, 0, 200, '2018-01-01', '2020-01-01']
    fetcher = ArgoDataFetcher(mode='expert', parallel=True, chunks='auto', box_maxsize=[10, 10, 50]).region(large_box)
    ds = fetcher.to_xarray()
    par_bench.append({'ETIM':time.time() - start_time, 'NPTS':np.max(ds['N_POINTS'].values), 
'MBYT': ds.nbytes/1e6, 'CHUNKS': fetcher.fetcher.chunks, 'NREQ': len(fetcher.fetcher.urls)})
    print("Run #%i" % run, par_bench[-1])
```

ps: The request fails for 0-400m and 0-800m boxes.

**Results:**
dev-pr28-erddap-chunks-bench-results-01

File: argopy/data_fetchers/erddap_data.py [hunk 1]
```diff
@@ -64,269 +66,374 @@ def cname(self):
...
+            "doxy_adjusted_error": np.float64,
+            "ptmp": np.float64,
+        }
         plist = self._minimal_vlist
         response = {}
         for p in plist:
             response[p] = dref[p]
         return response
 
+    def cname(self):
```
Comment (quai20) [inline argopy/data_fetchers/erddap_data.py:346]: Do we still need the abstract method **cname** now that this one exists ?

**Review by quai20 — APPROVED** [review]
I'll continue testing, but code is ok to me, and based on [](https://argopy.readthedocs.io/en/parallel-requests/performances.html#Parallel-data-fetching) everything works fine.

**Comment by gmaze** [comment]
ok, do you think, this is ready for a merge ?

**Comment by quai20** [comment]
> ok, thanks @quai20 for the review
> do you think, this is ready for a merge ?

ok to me !

**Pull request merged**